clang/test/SemaCXX/cxx2c-pack-indexing.cpp

```diff
@@ -160,3 +160,22 @@ namespace GH88929 {
     using E = P...[0]; // expected-error {{unknown type name 'P'}} \
                        // expected-error {{expected ';' after alias declaration}}
 }
+
+namespace GH88925 {
+template <typename...> struct S {};
+
+template <int...> struct sequence {};
+
+template <typename... Args, int... indices> auto f(sequence<indices...>) {
+  return S<Args...[indices]...>(); // #use
+}
+
```

Comment:
I suspect there is the same problem for expressions

```cpp
template <auto...> struct S2 {};
template <auto... Args, int... indices> auto f(sequence<indices...>) {
  return S2<Args...[indices]...>(); // #use
}
```

Reply:
Oops, yeah, good catch! Honestly, I have tried one similar example (with a fold expression) but had no luck:
```cpp
template <auto... Args, int... indices> auto f(sequence<indices...>) {
  return (Args...[indices] + ...);
}
```

Let me see what's happening here.

Reply:
> ```cpp
> template <auto...> struct S2 {};
> template <auto... Args, int... indices> auto f(sequence<indices...>) {
>  return S2<Args...[indices]...>(); // #use
> }
> ```

OK, so the problem here is *not* the same: we end up [classifying the `PackIndexingExpr`'s category](https://github.com/llvm/llvm-project/blob/338561657685c1831a53563b1bc36ffc7470239e/clang/lib/AST/ExprClassification.cpp#L219-L220) while deducing against the NTTP of `S2`. This happens before the instantiation of `f`, and therefore `Args...[indices]...` is still dependent.

I *guess* we can perceive dependent `PackIndexingExprs` as LValues in such cases, like what we did for `UnresolvedLookupExpr`. Additionally, P2662 now limits them to id-expressions only, and hence we can probably always treat them as LValues?

Reply:
> I guess we can perceive dependent PackIndexingExprs as LValues in such cases, like what we did for UnresolvedLookupExpr. Additionally, P2662 now limits them to id-expressions only, and hence we can probably always treat them as LValues?

Yes, they are always LValues for now.  So we might want to do that indeed

Reply:
I realized those expressions could be PRValues after an instantiation because in `PackIndexingExpr` we have:
```cpp
if (!isInstantiationDependent())
  setValueKind(getSelectedExpr()->getValueKind());
```

So, I'm going to make it always an LValue iff it's dependent.